### DISCUSSION TITLE
Documentation generation

### DISCUSSION THREAD

**Discussion by applejag** [deepnoodle-ai/risor#168] [Ideas]
Would be nice with a set convention on how to document Risor code.

Preferably in a way that can be parsed, to eventually become the input for generating the docs on the website.

Some wanted use cases for function docs:

- Define signature
- Define description
- Define parameter descriptions
- Define examples
- Parsable syntax, so it can be used in documentation generation and language servers
- No markdown parsing should be needed

Some inspiration sources:

- Rust's `///` documentation comments: https://doc.rust-lang.org/book/ch14-02-publishing-to-crates-io.html#making-useful-documentation-comments
- JavaScript's `/** @class */` JSDoc: https://jsdoc.app/
- LUA's `---@class` EmmyLua: https://emmylua.github.io/
- C#'s `/// <summary>` documentation comments: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/documentation-comments

This was previously discussed in https://github.com/risor-io/risor/issues/118

Should Risor add a type syntax, similar to Python and TypeScript? Or should we solely rely on comments like JSDoc and EmmyLua?

Let's discuss different approaches.

**Comment by applejag** [comment]
I strongly think that the generated documentation should only be parsed from Risor code, so that end users also could make use of documentation generation (such as in generating Markdown for company internal modules written in Risor).

This might need some new Risor syntax to be able to accommodate the documentation style used, such as the spread operator `...` for variadic parameters.

I also suggest adding a syntax for declaring functions, to only be used inside the Risor builtin modules, by using a semicolon `;` instead of curly braces `{ }`.

Example:

```go
// file: modules/strings/strings.risor

// Returns true if the string s contains substr.
// @param {string} s The string to search.
// @param {string} substr The substring to try to find inside `s`.
// @return {bool}
//
// @example
// >>> strings.contains("abc", "b")
// true
// >>> strings.contains("abc", "d")
// false
// >>> "abc" | strings.contains("b")
// true
func contains(s, substr);

// Repeat returns a new string consisting of `count` copies of the string `s`.
// @param {string} s The string to repeat.
// @param {int} count The number of times to repeat `s`.
// @return {string}
//
// @example
// >>> strings.repeat("na", 5) + " batman"
// "nanananana batman"
func repeat(s, count);
```

I suggest borrowing the JSDoc syntax, as some developers will find it familiar. And it's a quite decent documentation convention.

For the generated parts from `cmd/risor-modgen` (introduced in #159), it could generate Risor code based on comments in the Go code. Such as:

```go
// file: modules/strings/strings.go

// Returns true if the string s contains substr.
//
// # Example
//
//   >>> strings.contains("abc", "b")
//   true
//   >>> strings.contains("abc", "d")
//   false
//   >>> "abc" | strings.contains("b")
//   true
//
//risor:generate
func contains(s, substr string) bool {
  // ...
}
```

The above Go snippet would generate the above Risor snippet, but to `modules/strings/strings_gen.risor` instead. It's making use of a [Go doc style header](https://tip.golang.org/doc/comment#headings) and a code block to declare the example.

Not sure how to annotate the parameters though. Maybe like this?

```go
// file: modules/strings/strings.go

func contains(
  s, // The string to search.
  substr string, // The substring to try to find inside `s`.
) bool {
  // ...
}
```

**Reply by applejag** [reply]
On second thoughts, it shouldn't matter if a function is generated via risor-modgen or not. Both should just use a manually written `strings.risor`. No `strings_gen.risor` file.

**Comment by applejag** [comment]
An alternative to instead of a JSDoc-esque approach is to add syntax for this.

Such as if you could add typing information that's only considered by your editor and potentially the compiler, but that is effectively ignored by the runtime. Similar to Python's type notation.

So similar approach to [my other comment](https://github.com/risor-io/risor/discussions/168#discussioncomment-8222725) with a stub Risor file, but using dedicated syntax for the type info:

```go
// file: modules/strings/strings.risor

// Returns true if the string s contains substr.
func contains(s, substr string) bool;

// Repeat returns a new string consisting of `count` copies of the string `s`.
func repeat(s string, count int) string;
```

One drawback is that we don't get a way to add descriptions to fields, but that's maybe an OK compromise.

Another drawback is that it's quite unclear how to declare examples then.

To continue with the thought of "add more syntax", then there could be an `example` keyword:

```go
// The "contains" part signals which symbol the example belongs to
//      ↓↓↓↓↓↓↓↓
example contains {
  >>> strings.contains("abc", "b")
  true
  >>> strings.contains("abc", "d")
  false
  >>> "abc" | strings.contains("b")
  true
}
```

The Go language uses functions prefixed with `Example` in a `*_test.go` file. But as Risor is a scripting language that evaluates from top to bottom (meaning the order of function declaration matters), and that I think Risor should not require multiple files, then adopting Go's approach does not look that well suited. Also, defining Risor's `>>>` style of examples is very difficult to shoehorn into Go's Example tests with their `// Outputs:` comment.

**Comment by luisdavim** [comment]
Yeah, I was thinking 3 options:

1. Comments on go files, possibly with some custom tags to capture specific properties like parameters, summary and things like that.
We can also search through the files in the examples folder and automatically embed the ones that call the function/mode in question into the generated MD file.

2. Add some fields to the module and built-in types to hold the documentation details, similar to cobra's long and short but instead of a free text field it could be a struct to capture specific properties like parameters, summary and whatnot. 
Here we can still search through the examples and include them automatically 

3. Add comments to the risor files in the examples folder and use those comments to generate the docs, something like godoc.

Or some combination of the above. 
I think what you are proposing is closer to no. 3 right?

**Reply by applejag** [reply]
1. I believe it's a good idea to decouple the docs gen from the Risor Go code. I find it important that end users can also document their own code, and it'd be good to have the docs from Risor builtins and docs from new Risor code be defined in the same way. So there's only 1 single way to document Risor code.

2. I think not, as documentation information is not worth keeping in the runtime. That's debug info. The `risor` binary should not contain those strings. Components like a language server or the docs generator could embed this, but not the main binary. It's just bloat, that has no effective need.

3. Yea that's what I'm thinking, but instead keep that code in the modules folder, right next to the source code (where the markdown files are right now since #157).

The benefit of doing docs from the Risor code in a parsable manner is that future tools can make use of it.

What I imagine is something like this:

````console
$ risor doc myfile.risor
# myfile

## Functions

### myfunc

```risor
func myfunc(a string, b int) bool
```

Some docs for myfunc

```risor
>>> myfunc("hello", 123)
true
```
````

^to generate some kind of simple markdown that you can store in your repo next to your Risor files.

But then if you want a different way to show docs, then you can output the docs as JSON and mold the docs as however you want, such as:

```console
$ risor doc myfile.risor --output json
{
  "file": "myfile.risor",
  "functions": {
    "myfunc": {
      "parameters": [
        {
          "name": "a",
          "type": "string"
        },
        {
          "name": "b",
          "type": "int"
        }
      ],
      "return": {
        "type": "bool"
      },
      "description": "Some docs for myfunc",
      "examples": [
        {
          "input": "myfunc(\"hello\", 123)",
          "output": "true"
        }
      ]
    }
  }
}
```

Then for the Risor official docs, it could have its own markdown generation that makes use of more Nextra features.

This adds more incremental steps, which may increase the complexity a little, but it allows a composable design.

**Reply by luisdavim** [reply]
> I believe it's a good idea to decouple the docs gen from the Risor Go code. I find it important that end users can also document their own code, and it'd be good to have the docs from Risor builtins and docs from new Risor code be defined in the same way. So there's only 1 single way to document Risor code.

I agree but I think these are separate concerns, one is how to document the language itself and the other is how to document code written in it, I think we can tackle both but should focus first on the way to automate the maintenance of the docs that go into https://risor.io/ and that comes mostly from the go code in this repo.

But I'm fine with any of the 3 options and I also agree the second one is the least desirable. 

Option 3 also forces the creation of examples which I think it a plus actually :D

**Reply by luisdavim** [reply]
BTW, regarding where we place the `.md` files I don't have a strong opinion and I'm ok with both placing it in the module folder as well as in some docs folder but placing them on the module folder does have the benefit that when you're browsing through the code in GH those are visible and that can be good for new users and contributors exploring the project.

**Reply by applejag** [reply]
Yea reminders of the focus is always good. Docs needs to go on risor.io first and foremost.

In its current state, docs are very prone to becoming out of sync. To optimize for reducing docs drift would be to have the docs as Go doc comments in the Go code. As close as possible to the implementation. And with the signatures automatically generated.

Another way to avoid drift would be to improve tooling. To have automated checks. Like a failing GitHub Action if a function is missing docs.

Also need to be careful to not lose control over the docs and only rely on the docs generation. Reference documentation should be as consistent as possible, but we don't want to lose the ability to add a custom MDX section to each module. A simple approach could be that in the `risor-site` repo each module docs page is just this:

```mdx
# gha

<GeneratedModuleDocs module="gha" />
```

So that you can add a big usage section like:

````mdx
import { Callout } from 'nextra/components';

# gha

## Usage

This is a longer description of how to use the module as a whole.

Lorem ipsum blah blah

```yaml
steps:
  - shell: risor {0}
    run: |
      if !yaml.valid(cat(".yamllint.yaml")) {
        gha.log_warning("File is not properly YAML formatted", {"file": ".yamllint.yaml"})
      }
```

<Callout type="info" emoji="ℹ️">
  Some info text
</Callout>

Some more explanations blah blah

<GeneratedModuleDocs module="gha" />
````

The above could maybe be achieved by defining a way to do module-level docs generation, but making use of more advanced MDX/Nextra features would become super tedious then, and it's not even portable (e.g GitHub doesn't know how to handle `<Callout>`).

How I see it is that `<GeneratedModuleDocs>` should be a custom component in `risor-site` repo that reads the generated docs from a JSON file (however that JSON file is produced, e.g via `risor docs --output json`) and then produces the wanted HTML out of it. All from inside the `risor-site` repo. All risor.io-specific styling is happily done in the Node project where you get all the intellisense help you need, and the "parsing Risor/Go code" part stays in this `risor` repo.

The only docs coming from this `risor` repo would be reference documentation, and very simple examples for some functions.

Then in the `risor-site` repo you add the other kinds of usage documentation, guides, etc.

As a first step, the `risor` CLI does not need to know how to generate Markdown. It should only care about generating these JSON files of the reference docs. At a later stage it could get its own markdown generator as a follow-up just for the convenience of the end users wanting to document their own code.

Hmm the more I think about it, the less sure I am about using Risor files as the source of the reference documentation.

---

Excuse my rambling. I'm thinking and writing my comment at the same time.

**Reply by luisdavim** [reply]
For additional manually crafted docs we can allow creating something like a `header.md` and/or `footer.md` that would be merged with the docs when publishing to the site. And that doesn't mean they are the actual header and footer on the final site, there can be another layer for that, it's just how they would be added before or after the generated ones.

**Reply by luisdavim** [reply]
> As a first step, the risor CLI does not need to know how to generate Markdown. It should only care about generating these JSON files of the reference docs. At a later stage it could get its own markdown generator as a follow-up just for the convenience of the end users wanting to document their own code.

I'm not sure about this, seems a bit extra, we'd need to generate json from comments and then markdown from json, don’t see much benefit in that.

Maybe if we parse the comments into some struct and this comes for free... But not if that adds complexity, we don't need this for generating the risor docs, and it seems like it would be more useful for the generation of docs for code written in risor.... I really think these are different problems, and we can have different solutions. If we can solve both with the same solution, awesome, but let's not make that a goal.
